fish-rust/src/bitset.rs

```diff
@@ -0,0 +1,275 @@
+use num_traits::bounds::UpperBounded;
+use num_traits::{AsPrimitive, FromPrimitive, Num};
+use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, Not, Shl, Shr};
+
+pub struct BitSet<T>(T);
+
+impl<
+        T: BitAnd<T, Output = T>
+            + BitOr<T, Output = T>
+            + BitAndAssign<T>
+            + BitOrAssign<T>
+            + Shl<usize, Output = T>
+            + Shr<usize, Output = T>
+            + Not<Output = T>
+            + AsPrimitive<u8>
+            + FromPrimitive
+            + UpperBounded
+            + Num
```

Comment:
Is there no Integral trait that implies the bit operations?

Reply:
Unfortunately not. I think [PrimInt](https://docs.rs/num-traits/latest/num_traits/int/trait.PrimInt.html) is the closest you get.